app/scripts/controllers/permissions/permissionsLog.js

```diff
@@ -0,0 +1,306 @@
+import {
+  CAVEAT_NAMES,
+  HISTORY_STORE_KEY,
+  LOG_STORE_KEY,
+  LOG_IGNORE_METHODS,
+  WALLET_PREFIX,
+} from './enums'
+
+const LOG_LIMIT = 100
+
+const getAccountToTimeMap = (accounts, time) => accounts.reduce(
+  (acc, account) => ({ ...acc, [account]: time }), {}
+)
+
+/**
+ * Create middleware for logging requests and responses to restricted and
+ * permissions-related methods.
+ */
+export default class PermissionsLogController {
+
+  constructor ({ restrictedMethods, store }) {
+    this.restrictedMethods = restrictedMethods
+    this.store = store
+  }
+
+  getLogStore () {
+    return this.store.getState()[LOG_STORE_KEY] || []
+  }
+
+  updateLogStore (logs) {
+    this.store.updateState({ [LOG_STORE_KEY]: logs })
+  }
+
+  getHistoryStore () {
+    return this.store.getState()[HISTORY_STORE_KEY] || {}
+  }
+
+  updateHistoryStore (history) {
+    this.store.updateState({ [HISTORY_STORE_KEY]: history })
+  }
+
+  createMiddleware () {
+    return (req, res, next, _end) => {
+
+      let activityEntry, requestedMethods
+      const { origin, method } = req
+      const isInternal = method.startsWith(WALLET_PREFIX)
+
+      // we only log certain methods
+      if (
+        !LOG_IGNORE_METHODS.includes(method) &&
+        (isInternal || this.restrictedMethods.includes(method))
+      ) {
+
+        activityEntry = this.logActivity(req, isInternal)
+
+        if (method === `${WALLET_PREFIX}requestPermissions`) {
+          // get the corresponding methods from the requested permissions
+          requestedMethods = this.getRequestedMethods(req)
+        }
+      } else if (method === 'eth_requestAccounts') {
+
+        // eth_requestAccounts is a special case; we need to extract the accounts
+        // from it
+        activityEntry = this.logActivity(req, isInternal)
+        requestedMethods = [ 'eth_accounts' ]
+      } else {
+        // no-op
+        return next()
+      }
+
+      // call next with a return handler for capturing the response
+      next(cb => {
+
+        const time = Date.now()
+        this.addResponse(activityEntry, res, time)
+
+        if (!res.error && requestedMethods) {
+          // any permissions or accounts changes will be recorded on the response,
+          // so we only log permissions history here
+          this.logPermissionsHistory(
+            requestedMethods, origin, res.result, time,
+            method === 'eth_requestAccounts',
+          )
+        }
+        cb()
+      })
+    }
+  }
+
+  // creates and commits an activity log entry, without response data
+  logActivity (request, isInternal) {
+    const activityEntry = {
+      id: request.id,
+      method: request.method,
+      methodType: isInternal ? 'internal' : 'restricted',
+      origin: request.origin,
+      request: cloneObj(request),
+      requestTime: Date.now(),
+      response: null,
+      responseTime: null,
+      success: null,
+    }
+    this.commitActivity(activityEntry)
+    return activityEntry
+  }
+
```

Comment:
I am going to make a comment that is a nitpick, and is also a nitpick that I frequently guilty of. I think the ordering of methods here can be made a bit more consistent.

For example, `addResponse` is defined after its use within this file, but `getAccountsFromPermission` is defined before its use within this file.

Reply:
Most recent commit includes a cleanup of this file, including updated method order, comprehensive comments for all class methods, and a fix for activity log responses never being written to the store.